### PR TITLE
support MTP with chunked loss

### PR DESCRIPTION
**Pull request by acisseJZhong**
## Summary

Enable DeepSeek-V3 MTP with chunked loss without materializing full-sequence
vocabulary logits for any model output.

`MTPDecoder.preprocess_inputs()` builds aligned tuples of main/MTP token inputs
and labels. `MTPDecoder` returns predictions in the same order, and `MTPLoss`
applies `mtp_scale / num_mtp_layers` to each auxiliary objective.
`ChunkedLossWrapper` chunks the aligned tuples and passes the same structure to
the inner loss, without depending on MTP-specific loss behavior.

## Design

- **Prepare aligned MTP pairs before sharding.** `MTPDecoder.preprocess_inputs()`
  adds shifted token, label, and validity-mask tensors and their input layouts
  before applying mask construction, CP sharding, and SPMD annotation. This
  preserves packed-document boundaries and the required roll-before-CP-shard
  ordering without changing the shared `Decoder` preprocessing API.
- **Keep `ChunkedLossWrapper` loss-agnostic.** It accepts either one
  prediction/labels pair or aligned tuples, chunks them together, and forwards
  the same structure to any inner loss. Existing single-output and RL losses
  keep their original path.
- **Extend the chunk loop and autograd bridge to tuples.**
  `ChunkedLossWrapper` creates chunks and a `GradAccumulator` for each
  prediction, applies the LM head to every prediction at each chunk index, and
  calls the inner loss once. The bridge returns each assembled gradient to its
  matching model output. The FSDP LM-head lifecycle now spans the full
  output-by-chunk loop.
- **Integrate DeepSeek-V3 MTP.** `MTPDecoder` returns main and per-depth hidden
  states, while `MTPLoss` applies cross entropy to the aligned labels and weights
  each auxiliary objective by `mtp_scale / num_mtp_layers`. GraphTrainer and the
  debug MTP configuration support the same multi-output contract.

This PR does not enable MTP with pipeline or context parallelism.

## Tests

Targeted CPU tests:

```bash
pytest -q \
  tests/unit_tests/cpu/test_loss.py \
  tests/unit_tests/cpu/test_trainer.py \
  tests/unit_tests/cpu/test_validate.py \
  torchtitan/experiments/graph_trainer/tests/test_chunked_loss.py \
  torchtitan/experiments/rl/tests/test_dapo_math.py
```

Result: `62 passed, 2 subtests passed`. This covers MTP loss and input
preprocessing, GraphTrainer chunked loss, and chunked/non-chunked DAPO parity.

GPU training also passed with MTP enabled for 1 GPU, 2-GPU TP/SP, and 4-GPU
FSDP configurations.

The single-output path was bitwise identical to the parent on CPU and H100 for
loss, hidden-state gradients, and LM-head gradients (`torch.equal=True`, maximum
absolute difference `0.0`).

For numerical validation, parent `ac10ca48f` and this change (`9edf88e16`) ran
for 5 deterministic training steps from the same seed checkpoint, using
`seed=42`, a 256-token context, and disabled CUDA graphs. Each configuration was
run twice; its full-precision TensorBoard loss and gradient-norm values were
identical across repeats.

| Configuration | Steps | Maximum loss relative difference | Maximum grad-norm relative difference |
| --- | ---: | ---: | ---: |
| 1 GPU | 5 | `6.41e-5` | `2.16e-4` |
| 4-GPU FSDP | 5 | `4.74e-5` | `4.18e-5` |

Both configurations completed successfully and followed the same loss and
gradient-norm trajectories. The small differences are expected from the changed
floating-point accumulation order introduced by chunking.

### CODE REVIEW

File: torchtitan/experiments/graph_trainer/chunked_loss.py [hunk 1]
```diff
@@ -67,15 +69,19 @@ class _ChunkedLossWrapperWithParamGrads(torch.autograd.Function):
...
+            raise ValueError(
+                "Graph chunked-loss autograd bridge expected at least "
+                f"{expected_fixed_args} arguments for {num_predictions} "
+                f"predictions, got {len(args)}."
+            )
+        accumulated_h_grads = args[num_predictions : 2 * num_predictions]
+        total_loss = args[2 * num_predictions]
+        lm_head = args[2 * num_predictions + 1]
+        fsdp_enabled = args[2 * num_predictions + 2]
+        lm_params = args[2 * num_predictions + 3 :]
```
Comment (tianyu-l) [inline torchtitan/experiments/graph_trainer/chunked_loss.py]: looks not very readable, are there better ways?

File: torchtitan/components/loss.py [hunk 2]
```diff
@@ -252,6 +270,30 @@ def _maybe_compile(self, compile_config: CompileConfig | None) -> None:
...
+    def _compute_loss_term(
+        self,
+        pred: torch.Tensor,
+        labels: torch.Tensor,
+        **loss_inputs: Any,
+    ) -> tuple[torch.Tensor, dict[str, torch.Tensor]]:
+        """Compute one unnormalized loss term."""
+        del loss_inputs
+        return self.fn(pred, labels), {}
+
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]: looks awkward that these are not consumed by non chunked `BaseLoss`?

File: torchtitan/components/loss.py [hunk 3]
```diff
@@ -309,6 +351,23 @@ def __call__(
             loss = loss / global_valid_tokens
         return loss, {}
 
+    def _compute_loss_term(
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]: same, this is only consumed by ChunkedLoss

File: torchtitan/models/deepseek_v3/mtp.py [hunk 4]
```diff
@@ -335,63 +335,95 @@ def __init__(self, config: Config, *, compile_config: CompileConfig | None = Non
...
-                mtp_labels,
-                global_vocab_size=self.global_vocab_size,
+            assert isinstance(mtp_labels, torch.Tensor)
+            loss_terms.append(
+                LossTerm(
+                    mtp_pred,
+                    mtp_labels,
+                    weight=mtp_weight,
+                )
             )
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/mtp.py]: It seems that the pattern here goes from "a for loop of label construction + loss computation" to "a for loop of constructing labels + another for-loop of loss computation". I guess the memory overhead of labels is small so it's OK?

File: torchtitan/models/deepseek_v3/mtp.py [hunk 5]
```diff
@@ -335,63 +335,95 @@ def __init__(self, config: Config, *, compile_config: CompileConfig | None = Non
...
-        assert mtp_loss is not None
-        if num_mtp_layers > 1:
-            mtp_loss = mtp_loss / num_mtp_layers
-        loss = main_loss + mtp_loss * self.mtp_scale
-        if global_valid_tokens is not None:
-            loss = loss / global_valid_tokens
-        return loss, {}
+        return tuple(loss_terms)
+
+    def _compute_loss_term(
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/mtp.py]: do we need this? Every overwrite looks the same to me

File: torchtitan/components/loss.py [hunk 6]
```diff
@@ -586,14 +658,25 @@ def __call__(
         """
         from torch.distributed._composable.fsdp import FSDPModule
 
-        hidden_states = pred
         num_chunks = self.num_chunks
         lm_head = self.lm_head
         assert lm_head is not None, "Set lm_head before calling ChunkedLossWrapper"
+        loss_terms = self.loss_fn._build_loss_terms(pred, labels, **loss_inputs)
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]: I wonder if this breaks RL losses https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/rl/losses/dapo.py

File: torchtitan/models/deepseek_v3/mtp.py [hunk 7]
```diff
@@ -335,63 +335,95 @@ def __init__(self, config: Config, *, compile_config: CompileConfig | None = Non
...
             )
+            weighted_loss = term_loss * loss_term.weight
+            loss = weighted_loss if loss is None else loss + weighted_loss
+        assert loss is not None
+        with spmd.no_typecheck():
+            if global_valid_tokens is not None:
+                loss = loss / global_valid_tokens
+        return loss, {}
+
+    def _build_loss_terms(
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/mtp.py]: Letting each base loss to have this term really feels like overfitting the entire design to MTP.

Is there a way we can build this outside in the MTP decoder, and not changing the loss & chunked loss protocol much? I feel we could do the `roll_mtp_sequence` on hidden states and labels in the MTP decoder, and send multiple MTP terms to the loss, with weight as extra `loss_inputs` which is preexisting for RL use cases.

**Review by tianyu-l — CHANGES_REQUESTED** [review]
would like to push for a more minimal design

File: torchtitan/components/loss.py (hunk 2 above)
Comment (acisseJZhong) [reply torchtitan/components/loss.py]: it will be used for both BaseLoss and CrossEntropyLoss. https://github.com/pytorch/torchtitan/blob/1682272f749848f82812dcf54ca0e5372e6704ef/torchtitan/components/loss.py#L783

File: torchtitan/models/deepseek_v3/mtp.py (hunk 4 above)
Comment (acisseJZhong) [reply torchtitan/models/deepseek_v3/mtp.py]: I feel we have to do this separation: MTP owns label alignment and weighting, while ChunkedLossWrapper owns chunked execution. Otherwise ChunkedLossWrapper needs to handle MTP logic.

File: torchtitan/models/deepseek_v3/mtp.py (hunk 5 above)
Comment (acisseJZhong) [reply torchtitan/models/deepseek_v3/mtp.py]: get rid of this by letting MTPLoss inheriting from CrossEntropyLoss.

File: torchtitan/trainer.py [hunk 8]
```diff
@@ -797,15 +746,17 @@ def forward_backward_step(
                 parallel_dims=self.parallel_dims,
                 parallelism=self.config.parallelism,
             )
-            self.ntokens_seen += labels.numel()
+            self.ntokens_seen += (
```
Comment (tianyu-l) [inline torchtitan/trainer.py:819]: add a note that this is for MTP, where index 0 has all the labels

File: torchtitan/models/deepseek_v3/mtp.py [hunk 9]
```diff
@@ -220,41 +217,117 @@ def __init__(self, config: Config):
...
+            *(batch.pop(f"mtp_input_tokens_{depth}") for depth in depths),
+        )
+        loss_labels = (
+            main_labels,
+            *(batch.pop(f"mtp_labels_{depth}") for depth in depths),
+        )
+        batch["mtp_input_valid_masks"] = tuple(
+            batch.pop(f"mtp_input_valid_mask_{depth}") for depth in depths
+        )
+        return input_tokens, loss_labels, batch
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/mtp.py:318]: I think `input_tokens` should be part of `batch` -- if you agree we can do it in separate PR.

Not sure if labels should be put in batch so we have a clean dict interface. I think we can do it in dataloader / batcher and pop "labels" out before sending to model forward.

File: torchtitan/models/deepseek_v3/mtp.py [hunk 10]
```diff
@@ -220,41 +217,117 @@ def __init__(self, config: Config):
                 )
             self.mtp_layers.append(layer_config.build())
 
+    def preprocess_inputs(
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/mtp.py:225]: after this refactor, are there any gaps for CP?

File: torchtitan/trainer.py [hunk 11]
```diff
@@ -829,33 +780,51 @@ def _forward_backward_body(
...
+                inputs, labels, extra_kwargs = cast(
+                    BaseModel, self.model_parts[0]
+                ).preprocess_inputs(
+                    {**input_dict, "labels": labels},
+                    parallel_dims=self.parallel_dims,
+                    parallelism=self.config.parallelism,
+                )
+                assert isinstance(inputs, torch.Tensor)
+                assert isinstance(labels, torch.Tensor)
+                self.ntokens_seen += labels.numel()
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: this is because MTP + PP doesn't work yet, right?

File: torchtitan/components/loss.py [hunk 12]
```diff
@@ -673,48 +704,63 @@ def _chunk(t):
...
+                            h_chunks, grad_accumulators, strict=True
+                        ):
+                            assert h_chunk.grad is not None
+                            grad_accumulator.add(h_chunk.grad)
+                            h_chunk.grad = None
 
             if fsdp_enabled:
                 lm_head.set_reshard_after_forward(True)
                 lm_head.set_reshard_after_backward(True)
-                lm_head.set_requires_gradient_sync(True, recurse=False)
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:706]: why this change?

File: torchtitan/components/loss.py [hunk 13]
```diff
@@ -752,59 +798,74 @@ def _combine_chunk_metrics(
...
     ) -> torch.Tensor:
         """Return a differentiable loss via _DecoderOutputGradientBackProp.
         When ``.backward()`` is called (by the trainer or PP schedule),
         autograd calls ``_DecoderOutputGradientBackProp.backward`` which
-        returns ``accumulated_grad`` as the gradient for ``hidden_states``,
+        returns each accumulated gradient for its corresponding hidden state,
         propagating through the decoder. Subclasses override to swap in a
         different autograd Function.
         """
+        del lm_head, fsdp_enabled
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]: unrelated: why sending these in when they are not used? are they consumed by graph trainer overwrite?

File: torchtitan/components/loss.py
Comment (tianyu-l) [inline torchtitan/components/loss.py:1]: Didn't review too carefully, but seems just handling tuples of inputs pred and labels. Sounds fine as long as numerics are bitwise on single pred label inputs.

File: torchtitan/models/deepseek_v3/mtp.py (hunk 9 above)
Comment (acisseJZhong) [reply torchtitan/models/deepseek_v3/mtp.py:318]: I think returning inputs, labels, and batch is clear enough. https://github.com/pytorch/torchtitan/blob/c215a12ee6945cd2c786cd1da864e4ea44a60123/torchtitan/trainer.py#L793  

are you saying we just return one batch dict? and pass `batch[inputs], batch[labels]`, and `batch` instead? Why is this cleaner?

File: torchtitan/models/deepseek_v3/mtp.py (hunk 10 above)
Comment (acisseJZhong) [reply torchtitan/models/deepseek_v3/mtp.py:225]: yeah I think CP should work, WIP in next stacked PR.

File: torchtitan/trainer.py (hunk 11 above)
Comment (acisseJZhong) [reply torchtitan/trainer.py]: yes. deleted this change/assertion after rebasing Simon's change.

File: torchtitan/components/loss.py (hunk 12 above)
Comment (acisseJZhong) [reply torchtitan/components/loss.py:706]: I think this is a no-op. We are setting gradient sync to True everytime for the last chunk anyways.

https://github.com/pytorch/torchtitan/blob/c215a12ee6945cd2c786cd1da864e4ea44a60123/torchtitan/components/loss.py#L679

File: torchtitan/components/loss.py (hunk 13 above)
Comment (acisseJZhong) [reply torchtitan/components/loss.py]: yes needed by graph trainer https://github.com/pytorch/torchtitan/blob/c215a12ee6945cd2c786cd1da864e4ea44a60123/torchtitan/experiments/graph_trainer/chunked_loss.py#L39

I changed this function from static method to instance method and let graph trainer override.

File: torchtitan/components/loss.py
Comment (acisseJZhong) [reply torchtitan/components/loss.py:1]: yeah tested, added to the test section.

File: torchtitan/models/deepseek_v3/mtp.py (hunk 9 above)
Comment (tianyu-l) [reply torchtitan/models/deepseek_v3/mtp.py:318]: maybe the output can stay separated because they are closer to model interface, but can we at least change the input of this function, and the output of batch_generator to merge labels into the dict?
https://github.com/pytorch/torchtitan/blob/main/torchtitan/trainer.py#L757

This may be a change on dataloader side, can be done in separate PR.

**Review by tianyu-l — APPROVED** [review]

File: torchtitan/models/common/decoder.py [hunk 14]
```diff
@@ -317,8 +318,7 @@ def preprocess_inputs(
         *,
         parallel_dims: ParallelDims,
         parallelism: ParallelismConfig,
-        max_num_documents: int | None = None,
-        max_context_length: int | None = None,
+        input_sharding: dict[str, SpmdType] | None = None,
```
Comment (tianyu-l) [inline torchtitan/models/common/decoder.py]: wait, why this change? seems not used in trainer, is it leftover from previous versions?

Comment (acisseJZhong) [reply torchtitan/models/common/decoder.py]: I need to provide this input_sharding because now it might contain non standard input keys. https://github.com/pytorch/torchtitan/blob/4a797f3d699f64e7d58372428c8313c52967498c/torchtitan/models/deepseek_v3/mtp.py#L266

Comment (tianyu-l) [reply torchtitan/models/common/decoder.py]: oh I see what's going on. `input_sharding` is created just for MTP(Decoder) to use via `super().preprocess_input()`, which sounds too hacky and fragile.

For now we could just copy whatever `Decoder.preprocess_input` into MTP and not call super() at all, the redundancy is not that much.

Comment (acisseJZhong) [reply torchtitan/models/common/decoder.py]: codex help me added an instance method _get_input_sharding, which Decoder just returns decoder_input_sharding(), but MTPDecoder will customize.

Comment (tianyu-l) [reply torchtitan/models/common/decoder.py]: sounds over-engineering for this single MTP case
